accepted/2023/preview-apis/preview-apis.md

```diff
@@ -0,0 +1,279 @@
+
+> error ML123: FancyLoggerExtensions.AddFancyLogging() is not a stable API. In
+  order to consume this experimental API, suppress this error from the call
+  site.
+
+David is happy to accommodate future breaking changes around the fancy logging
+feature. He understands that the diagnostic ID is specific to the fancy logging
+feature of Martin's library, so he decides to add a suppression to the project
+file:
+
+```xml
+<Project>
+    <PropertyGroup>
+        <!-- Fancy logging is an experimental APIs -->
+        <NoWarn>$(NoWarn);ML123</NoWarn>
+    </PropertyGroup>
+</Project>
+```
+
+### Consuming another experimental API
+
+Later, David discovers Martin's library also has support for fancy serialization.
+When he starts consuming it, he gets another compilation error:
+
+> error ML456: FancySerializationExtensions.AddFancySerialization() is not a
+  stable API. In order to consume this experimental API, suppress this error
+  from the call site.
+
+Giving it some thought, David concludes that the risk of using an unstable
+serialization API isn't worth it because it's not necessary for Jabber, so he
+stops exploring that API.
+
+## Requirements
+
+### Goals
+
+* Don't change the guardrails for platform-level preview features:
+    - They require opt-in (without, the build fails with an error)
+    - They require to be viral (i.e. opt-in is also required for indirect
+      dependencies relying on preview features)
+    - The opt-in can be expressed centrally in the project file
+    - They require SDK and framework to match
+
+* Offer library-level preview features with a different set of guardrails:
+    - They require opt-in (without, the build fails with an error)
+    - The opt-in isn't viral and can be an implementation detail
+    - The opt-in is centralized, but it's specific to each preview feature
+    - Using them poses no requirements on the SDK version
+
+### Non-Goals
+
+* Relaxing or extending the guardrails for platform-level preview features
+
+## Stakeholders and Reviewers
+
+* Runtime teams
+* C#, VB, F# compiler and IDE teams
+* Library teams
+    - Core libraries
+    - ASP.NET
+    - EF
+    - Azure SDK
+    - R9
+* Project system team
+* SDK team
+* MSBuild team
+* NuGet team
+
+## Design
+
+### Attribute
+
+We'll create a new attribute, separate from the existing `System.Runtime.Versioning.RequiresPreviewFeaturesAttribute`:
+
+```C#
+namespace System.Diagnostics.CodeAnalysis;
+
+[AttributeUsage(AttributeTargets.Assembly |
+                AttributeTargets.Module |
+                AttributeTargets.Class |
+                AttributeTargets.Struct |
+                AttributeTargets.Enum |
+                AttributeTargets.Constructor |
+                AttributeTargets.Method |
+                AttributeTargets.Property |
+                AttributeTargets.Field |
+                AttributeTargets.Event |
+                AttributeTargets.Interface |
+                AttributeTargets.Delegate, Inherited = false)]
```

Comment:
Is this missing `AttributeTargets.Parameter` intentionally ?

Reply:
Yes. I don't know what it would mean to only mark a parameter as experimental as removing/changing a parameter is equivalent of removing the method and offering a new method with a different set of parameters.